robolectric/src/test/java/org/robolectric/shadows/ShadowParcelFileDescriptorTest.java

```diff
@@ -55,6 +60,174 @@ public void testOpens() throws Exception {
+    assertThat(pfd).isNotNull();
+    assertThat(pfd.getFileDescriptor().valid()).isTrue();
+
+    int fd = pfd.getFd();
+    assertThat(fd).isGreaterThan(0);
+
+    assertThat(pfd.detachFd()).isEqualTo(fd);
+
+    ParcelFileDescriptor newPfd = ParcelFileDescriptor.adoptFd(fd);
+
+    assertThat(newPfd).isNotNull();
+    assertThat(newPfd.getFileDescriptor().valid()).isTrue();
+
+    FileOutputStream os = new FileOutputStream(newPfd.getFileDescriptor());
+    os.write(5);
+    os.close();
+    newPfd.close();
+  }
+
+  @Test
+  public void testAdoptFd_canNotWriteReadOnlyFile() throws Exception {
+    assumeThat(TIPS_TESTS_NOT_SUPPOSED_TO_BE_RUN_ON_WINDOWS, File.separator, Matchers.equalTo("/"));
+
+    pfd = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY);
+    assertThat(pfd).isNotNull();
+    assertThat(pfd.getFileDescriptor().valid()).isTrue();
+
+    int fd = pfd.getFd();
+    assertThat(fd).isGreaterThan(0);
+
+    assertThat(pfd.detachFd()).isEqualTo(fd);
+
+    ParcelFileDescriptor newPfd = ParcelFileDescriptor.adoptFd(fd);
+
+    assertThat(newPfd).isNotNull();
+    assertThat(newPfd.getFileDescriptor().valid()).isTrue();
+
+    try (FileOutputStream os = new FileOutputStream(newPfd.getFileDescriptor())) {
+      assertThrows(IOException.class, () -> os.write(5));
+    }
+    newPfd.close();
+  }
+
+  @Test
+  public void testDetachFd() throws Exception {
+    assumeThat(TIPS_TESTS_NOT_SUPPOSED_TO_BE_RUN_ON_WINDOWS, File.separator, Matchers.equalTo("/"));
+
+    pfd = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_WRITE);
+    assertThat(pfd).isNotNull();
+    assertThat(pfd.getFileDescriptor().valid()).isTrue();
+
+    int fd = pfd.getFd();
+    assertThat(fd).isGreaterThan(0);
+
+    int detachedFd = pfd.detachFd();
+
+    assertThat(detachedFd).isEqualTo(fd);
+    assertThrows(IllegalStateException.class, () -> pfd.getFd());
+  }
+
+  @Test
+  public void testDetachFd_alreadyClosed() throws Exception {
+    assumeThat(TIPS_TESTS_NOT_SUPPOSED_TO_BE_RUN_ON_WINDOWS, File.separator, Matchers.equalTo("/"));
+
+    pfd = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_WRITE);
+    assertThat(pfd).isNotNull();
+    assertThat(pfd.getFileDescriptor().valid()).isTrue();
+
+    pfd.close();
+
+    assertThrows(IllegalStateException.class, () -> pfd.detachFd());
+  }
+
+  @Test
+  public void testDetachFd_afterDetachFd() throws Exception {
+    assumeThat(TIPS_TESTS_NOT_SUPPOSED_TO_BE_RUN_ON_WINDOWS, File.separator, Matchers.equalTo("/"));
+
+    pfd = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_WRITE);
+    assertThat(pfd).isNotNull();
+    assertThat(pfd.getFileDescriptor().valid()).isTrue();
+
+    int fd = pfd.getFd();
+    assertThat(fd).isGreaterThan(0);
+
+    int detachedFd = pfd.detachFd();
+
+    assertThat(detachedFd).isEqualTo(fd);
+    assertThrows(IllegalStateException.class, () -> pfd.getFd());
```

Comment:
We don't need to test it here because `testGetFd_afterDetachFd` has covered this occasion.

Reply:
Done. Line 226 is removed. And some redundant assertions are removed as well.